[BUG] Bug in Eclipse when passing a ternary operator with method references to an extension method

**Describe the bug**
Trying to use a ternary operator with arguments that are method references, as parameter to an `@ExtensionMethod` did not work.

```
foobar.setPredicate(value?Bar::isBaz:null);
```

**To Reproduce**

```
public class FooExtensions<T> {}
	public static <T> void setPredicate(Foo<T> foo, SerializablePredicate<T> predicate) {}
}

public class Foo<T> {}

@Getter
@Setter
public class Bar {
  boolean baz;
}

@ExtensionMethod(FooExtensions.class)
public class Test {

  void setPredicate(Foo<Bar> foobar, boolean value) {
	foobar.setPredicate(value?Bar::isBaz:null);		
  }

}
```

Compiling with javac succeeds.

As a workaround for Eclipse, replacing the ternary expression with separate calls `foobar.setPredicate(Bar::isBaz);` and `foobar.setPredicate(null);` compiles fine (per #2741).


> If you have stack traces or error messages please include all of them
```
!MESSAGE Problems occurred when invoking code from plug-in: "org.eclipse.jdt.ui".
Java Model Exception: java.lang.NullPointerException

!ENTRY org.eclipse.jdt.ui 4 0 2022-03-28 16:50:09.417
!MESSAGE Error in JDT Core during reconcile
!STACK 1
Java Model Exception: java.lang.NullPointerException
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:326)
	at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:348)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.makeConsistent(ReconcileWorkingCopyOperation.java:193)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:94)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:740)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:806)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1319)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:131)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy$1.run(JavaReconcilingStrategy.java:93)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:90)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:157)
	at org.eclipse.jdt.internal.ui.text.CompositeReconcilingStrategy.reconcile(CompositeReconcilingStrategy.java:94)
	at org.eclipse.jdt.internal.ui.text.JavaCompositeReconcilingStrategy.reconcile(JavaCompositeReconcilingStrategy.java:107)
	at org.eclipse.jface.text.reconciler.MonoReconciler.process(MonoReconciler.java:76)
	at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:210)
Caused by: java.lang.NullPointerException
Caused by: java.lang.NullPointerException
!SUBENTRY 1 org.eclipse.jdt.core 4 1005 2022-03-28 16:50:09.417
!MESSAGE java.lang.NullPointerException
!STACK 0
java.lang.NullPointerException
```

**Expected behavior**

I expect `foobar.setPredicate(value?Bar::isBaz:null);` to compile when using Eclipse 2021-09.

**Version info (please complete the following information):**
 - Lombok version: 1.18.22
 - Platform: Eclipse 2021-09 (4.21.0)

--- a/src/eclipseAgent/lombok/eclipse/agent/PatchExtensionMethod.java
+++ b/src/eclipseAgent/lombok/eclipse/agent/PatchExtensionMethod.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2012-2021 The Project Lombok Authors.
+ * Copyright (C) 2012-2022 The Project Lombok Authors.
  * 
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
@@ -47,6 +47,7 @@ import org.eclipse.jdt.internal.compiler.ast.ASTNode;
 import org.eclipse.jdt.internal.compiler.ast.Annotation;
 import org.eclipse.jdt.internal.compiler.ast.ClassLiteralAccess;
 import org.eclipse.jdt.internal.compiler.ast.CompilationUnitDeclaration;
+import org.eclipse.jdt.internal.compiler.ast.ConditionalExpression;
 import org.eclipse.jdt.internal.compiler.ast.Expression;
 import org.eclipse.jdt.internal.compiler.ast.MessageSend;
 import org.eclipse.jdt.internal.compiler.ast.NameReference;
@@ -298,7 +299,7 @@ public class PatchExtensionMethod {
 				List<TypeBinding> argumentTypes = new ArrayList<TypeBinding>();
 				for (Expression argument : arguments) {
 					TypeBinding argumentType = argument.resolvedType;
-					if (argumentType == null && Reflection.isFunctionalExpression(argument)) {
+					if (argumentType == null && requiresPolyBinding(argument)) {
 						argumentType = Reflection.getPolyTypeBinding(argument);
 					}
 					if (argumentType == null) {
@@ -338,8 +339,8 @@ public class PatchExtensionMethod {
 						} else {
 							param = parameters[i];
 						}
-						// Resolve types for lambdas
-						if (Reflection.isFunctionalExpression(arg)) {
+						// Resolve types for polys
+						if (requiresPolyBinding(arg)) {
 							arg.setExpectedType(param);
 							arg.resolveType(scope);
 						}
@@ -377,6 +378,10 @@ public class PatchExtensionMethod {
 		MessageSend_postponedErrors.clear(methodCall);
 		return resolvedType;
 	}
+
+	private static boolean requiresPolyBinding(Expression argument) {
+		return Reflection.isFunctionalExpression(argument) || argument instanceof ConditionalExpression && Reflection.isPolyExpression(argument);
+	}
 	
 	private static NameReference createNameRef(TypeBinding typeBinding, ASTNode source) {
 		long p = ((long) source.sourceStart << 32) | source.sourceEnd;
@@ -407,6 +412,7 @@ public class PatchExtensionMethod {
 		public static final Field argumentTypes = Permit.permissiveGetField(MessageSend.class, "argumentTypes");
 		public static final Field argumentsHaveErrors = Permit.permissiveGetField(MessageSend.class, "argumentsHaveErrors");
 		public static final Field inferenceContexts = Permit.permissiveGetField(MessageSend.class, "inferenceContexts");
+		private static final Method isPolyExpression = Permit.permissiveGetMethod(Expression.class, "isPolyExpression");
 		private static final Class<?> functionalExpression;
 		private static final Constructor<?> polyTypeBindingConstructor;
 		
@@ -430,6 +436,16 @@ public class PatchExtensionMethod {
 		public static boolean isFunctionalExpression(Expression expression) {
 			if (functionalExpression == null) return false;
 			return functionalExpression.isInstance(expression);
+		}
+		
+		public static boolean isPolyExpression(Expression expression) {
+			if (isPolyExpression == null) return false;
+			try {
+				return (Boolean) isPolyExpression.invoke(expression);
+			} catch (Exception e) {
+				// Ignore
+			}
+			return false;
 		}
 		
 		public static TypeBinding getPolyTypeBinding(Expression expression) {

--- a/test/transform/resource/after-delombok/ExtensionMethodFunctional.java
+++ b/test/transform/resource/after-delombok/ExtensionMethodFunctional.java
@@ -11,6 +11,7 @@ class ExtensionMethodFunctional {
 		test = ExtensionMethodFunctional.Extensions.map(test, s -> ExtensionMethodFunctional.Extensions.reverse(s));
 		ExtensionMethodFunctional.Extensions.consume(test, s -> System.out.println("1: " + s), s -> System.out.println("2: " + s));
 		ExtensionMethodFunctional.Extensions.consume(test, System.out::println, System.out::println);
+		ExtensionMethodFunctional.Extensions.consume(test, test.length() > 0 ? System.out::println : null);
 		ExtensionMethodFunctional.Extensions.toList1(Stream.of("a", "b", "c").map(String::toUpperCase));
 		List<Integer> i2 = ExtensionMethodFunctional.Extensions.toList2(Stream.of("a", "b", "c").map(String::toUpperCase));
 	}

--- a/test/transform/resource/after-ecj/ExtensionMethodFunctional.java
+++ b/test/transform/resource/after-ecj/ExtensionMethodFunctional.java
@@ -36,6 +36,7 @@ import lombok.experimental.ExtensionMethod;
     test = ExtensionMethodFunctional.Extensions.map(test, (<no type> s) -> ExtensionMethodFunctional.Extensions.reverse(s));
     ExtensionMethodFunctional.Extensions.consume(test, (<no type> s) -> System.out.println(("1: " + s)), (<no type> s) -> System.out.println(("2: " + s)));
     ExtensionMethodFunctional.Extensions.consume(test, System.out::println, System.out::println);
+    ExtensionMethodFunctional.Extensions.consume(test, ((test.length() > 0) ? System.out::println : null));
     ExtensionMethodFunctional.Extensions.toList1(Stream.of("a", "b", "c").map(String::toUpperCase));
     List<Integer> i2 = ExtensionMethodFunctional.Extensions.toList2(Stream.of("a", "b", "c").map(String::toUpperCase));
   }

--- a/test/transform/resource/before/ExtensionMethodFunctional.java
+++ b/test/transform/resource/before/ExtensionMethodFunctional.java
@@ -16,6 +16,8 @@ class ExtensionMethodFunctional {
 		test.consume(s -> System.out.println("1: " + s), s -> System.out.println("2: " + s));
 		test.consume(System.out::println, System.out::println);
 		
+		test.consume(test.length() > 0 ? System.out::println : null);
+		
 		Stream.of("a", "b", "c").map(String::toUpperCase).toList1();
 		List<Integer> i2 = Stream.of("a", "b", "c").map(String::toUpperCase).toList2();
 	}

